### ISSUE TITLE
twisted.conch.endpoints._NewConnectionHelper leaks agent connection

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8138]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sarum90 reported|
|-|-|
|Trac ID|trac#8138|
|Type|defect|
|Created|2015-12-04 19:32:58Z|
|Branch|https://github.com/twisted/twisted/tree/conch-endpoints-newconnection-agent-8138|

If agentEndpoint is supplied to twisted.conch.endpoints._NewConnectionHelper, then it creates a transport in _CommandTransport.secureConnection that is never cleaned up.

I believe this is equivalent to claiming that the transport that is located at userauth.agent.transport does not have any calls to transport.loseConnection() in cleanup code.

Ideally, after calling:

```
c = _NewConnectionHelper_with_agentendpoint_instance.secureConnection()
_NewConnectionHelper_with_agentendpoint_instance.cleanupConnection(c, dont_care)
```

All connections associated with the agent endpoint would be cleaned up.

Attachments:

* [fix-leaking-agentendpoint-connections-8138-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/12b84559ddcf2cff053c9fad12fc73d7e16bbc90.patch) (5119 bytes) - added by sarum90 on 2015-12-04 19:38:24Z - 
* [fix-leaking-agentendpoint-connections-8138-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/a96174e2fb1b2b7425b16c71bb363a1045e9f86b.patch) (4819 bytes) - added by sarum90 on 2015-12-04 19:40:52Z - 
* [fix-leaking-agentendpoint-connections-8138-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/9abb5ace5266cb12b94528372dbf95cddc8fa624.patch) (5033 bytes) - added by sarum90 on 2015-12-05 02:29:18Z - 
* [conch-endpoints-newconnection-agent-8138-move-connect-to-agent.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/59d1160bb78870dfb5c7306680e0cd20a8bb22f7.patch) (2410 bytes) - added by sarum90 on 2015-12-07 18:27:41Z - 
* [conch-endpoints-newconnection-agent-8138-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/12e1365646a7746d2c761e425edcbe2a43774164.patch) (4924 bytes) - added by sarum90 on 2015-12-07 19:58:12Z - 
* [conch-endpoints-newconnection-agent-8138-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8e4/8e48f7be41e1f2be5257273ae7bd95eec94aacb8/1f0968e027e56101133e5c553450c822fb813c05.patch) (606 bytes) - added by sarum90 on 2015-12-08 01:22:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8138 8138
type__defect defect
reporter__sarum90 sarum90
priority__normal normal
milestone__None None
branch__branches_conch_endpoints_newconnection_agent_8138 branches/conch-endpoints-newconnection-agent-8138
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1449257578152898 1449257578152898
changetime__1450953326372973 1450953326372973
version__None None
owner__sarum90 sarum90
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|sarum90 removed owner|
|-|-|

Apologies if I've left any of the the fields on the ticket in an incorrect state, first time reporting a bug and fix to twisted.

Ignore the first patch (fix-leaking-agentendpoint-connections-8138-1.patch​), it has a silly mistake in it (I don't think there's a way to delete attachments?).

Second patch is my proposed fix for the leaking SSH agent connections (fix-leaking-agentendpoint-connections-8138-2.patch​).  I don't fully understand all the layering of the twisted conch code, so I'm suspect that I may be crossing layers in bad ways with this solution. I'm also not crazy about the new userauth member variable to _ConnectionReady and the circular references it introduces. 

I'd be happy to take any suggestions for how to improve the patch.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Thanks for your contribution.

I will try to take a detailed look later.

I can confirm that the ticket is ok and is marked as needing review :)

----

the new `userauth` public API exposes the private _UserAuth class... which make _UserAuth also public... if needed we should also rename _UserAuth to UserAuth as it will be now public. 

I hope that we can find a way to keep this part private.

-------

```
connection.userauth.agent.transport.loseConnection()
```

crosses a lot of layer... indeed :)

can we have connection._userauth as a private member and update connection.loseConnection() to take care closing the connection using its private state? 

Also, for the test, I think that there should be a single StringTransport() which is shared by all conch layers.


Thanks again for your contribution!

Will come back with more details later... if meanwhile it will not be reviewed by someone else

**Comment by twisted-trac** [comment]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|sarum90 commented|
|-|-|

Thanks for the review! I made some changes, hopefully for the better.

----

> the new userauth public API exposes the private _UserAuth class... which make _UserAuth also public... if needed we should also rename _UserAuth to UserAuth as it will be now public.
> 
> I hope that we can find a way to keep this part private.

I realized that the _UserAuth was more naturally "owned" by the _CommandTransport that created it, which removed that issue, everything is nice and private now.

----

>can we have connection._userauth as a private member and update connection.loseConnection() to take care closing the connection using its private state?

went for connectionLost() on the _CommandTransport Protocol rather than loseConnection... I think connectionLost() will always be called whereas loseConnection might only be called when the termination is occurring from this side?

----

>Also, for the test, I think that there should be a single StringTransport() which is shared by all conch layers.

I don't think this would be possible in this case? The reason being that the connection to the agent is a connection to a different Endpoint than the outgoing ssh connection (normally a unix socket I think?).  Either way, with the cleanup code now on a protocol that the API does not provide a reference to, I found this easier to test by extending the current agentEndpoint test at a higher level rather than this really low level unit test.

----

Thanks again for the quick first review! Looking forward to the more detailed one.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

(In [46429]) Branching to conch-endpoints-newconnection-agent-8138.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to sarum90|
|-|-|

Hi and thanks for the update!

_CommandTransport._userauth, even if is private, still need docstring just like a public API... you can try to reuse the removed docstring from _connectToAgent and inform that it is none if no secured connection was done yet.

--------

The patch needs a NEWS file fragment so that your change are advertised as part of the relese notes.

More info here 
http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

I have updated your patch to include a start fragment. Please review it and updated it as I put it there just to help you get a quick start.

-------


I think that
```
if self._userauth:
    if self._userauth.agent:
        doSomething()
```

can be rewritten to reduce the indentation

```
if self._userauth and self._userauth.agent:
    doSomething()
```


... but since we just want to lose the connection for _UserAuth can we have something like

```
if self._userauth:
    self._userauth.loseAgentConnection()
```

I hope this express more clearly why the code is there.

------

For the variables from the test please check the Twisted naming convention http://twistedmatrix.com/documents/current/core/development/policy/index.html

client_transport vs clientTransport.

but the IOPump should already contain a reference to the client transport.. so you can just use pump.clientIO


I now understand why we need 2 transports. One is for SSH client to server and another one is for client to agent. Is that correct?

-----

It would be nice to a dedicated test for this behaviour maybe named test_agentAuthenticationConnectionLost
The test's docstring can describe into details what is tests and what is the expected behavior... or you can extend the current docstring.

Please add a comment for each time you pump the connected transport.

I find it hard to read the current test ... and since we don't have tests for the test code it is very important to have easy to ready tests so that we can validate them just by reading them.

------

While trying to understand the test changes and reading to code I find it strange that we have to call _UserAuth.agent.transport.loseConnection()

I hope that the new loseAgentConnection is expression the intention of the code better.

----------

Please check my latest commit from the branch attached to this ticket and submit a patch based on it.

Please review my changes in the code and news file fragment.

Free free to revert any of my changes.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|sarum90 commented|
|-|-|

Replying to [adiroiban](#note_6):

I'm still playing around with the test a bit to figure out the right thing to do there, but I wanted to address:

> 
> While trying to understand the test changes and reading to code I find it strange that we have to call _UserAuth.agent.transport.loseConnection()
> 
> I hope that the new loseAgentConnection is expression the intention of the code better.
> 

If you think that increases readability, great! However, I'm not crazy about the fact that _UserAgent is then cleaning something up that was created by a different object . If we're adding loseAgentConnection, I think it follows that _connectToAgent should be moved into _UserAgent as well (for symmetry of creation and destruction code).  I've done that in the attached conch-endpoints-newconnection-agent-8138-move-connect-to-agent.patch​. If you think that's a good idea, then I think we should apply that to the svn branch.

I also don't like that you have to call agent.transport.loseConnection when you never have to call any creation code on agent.tranport prior to that, but that looks like a twisted-ism that is pretty ingrained in the code at this point.

keeping assigned to me until I supply a patch fixing up the tests as well.

**Comment by twisted-trac** [comment]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|sarum90 set owner to @adiroiban|
|-|-|

Sorry about the delay, my laptop ran out of power this weekend and I left my charger at work. 

Thanks again for the review. I feel like you are really helping this go from a minimal set of code that technically does fix the issue to something that is actually long-term maintainable and fixes the issue correctly.

Comments addressed in conch-endpoints-newconnection-agent-8138-4.patch and responded to below.

Replying to [adiroiban](#note_6):
> _CommandTransport._userauth, even if is private, still need docstring just like a public API... you can try to reuse the removed docstring from _connectToAgent and inform that it is none if no secured connection was done yet.

Added in patch.

> 
> --------
> I have updated your patch to include a start fragment. Please review it and updated it as I put it there just to help you get a quick start.

Reviewed your news file with the intent to update it, but what you have there already seems more succinct and accurate that anything I could come up with. Thanks for adding the file!

> 
> ------
> 
> [...]the IOPump should already contain a reference to the client transport.. so you can just use pump.clientIO
You are correct, it appears I didn't do quite enough code-grepping before sending out the test :). Thanks for fixing that!

> 
> I now understand why we need 2 transports. One is for SSH client to server and another one is for client to agent. Is that correct?
That is my understanding, yup.

> 
> -----
> 
> It would be nice to a dedicated test for this behaviour maybe named test_agentAuthenticationConnectionLost
> The test's docstring can describe into details what is tests and what is the expected behavior... or you can extend the current docstring.
Given that you cannot test cleanup without testing setup, it seems more natural to me to test the cleanup of the agent connection with the current test that actually sets up the agent connection. Thus I’ve biased toward just extending the current docstring.

I know that’s not everyone’s view on testing though. So if you really want separate tests for [Setup] and [(setup and) cleanup] I’ll be happy to refactor.

> 
> Please add a comment for each time you pump the connected transport.
Evidently I was pumping it 1 to many times! Thanks for the comment. I actually just copied the code from test_loseConnection. Rather than doing that I’ve added a new helper method that has all of the comments from test_loseConnection, and modified both tests to just call the function.

There is 1 oddity with this approach… with the helper method, the test_loseConnection no longer has any assertions in it. Again, I’m alright with this because I know that the new helper has validation, but there are testing philosophies that every test should have at least one (or perhaps exactly one) assertion. I’m unsure if Twisted has any requirements like this. If so let me know, preferably with a suggestion of how you would rather the tests be segmented.

> 
> I find it hard to read the current test ... and since we don't have tests for the test code it is very important to have easy to ready tests so that we can validate them just by reading them.
> 

So… confession time, I noticed this bug because I was using twisted in a different project and that project's tests fail if the reactor is unclean. The reason I have confidence in this fix is because with this patch the tests in that project no longer have an unclean reactor when they are run.  I don’t actually gain that much confidence from the added unit test for the reasons you’ve outlined above (how are we sure that what we are doing in the test is representative of what a real transport does?). I’d definitely like to have enough confidence and readability in these tests so that both you and I can be convinced of correctness of the fix from the tests in twisted.conch.test.* alone.

Hopefully the new arrangement of code is better.

> ------
> 
> Please check my latest commit from the branch attached to this ticket and submit a patch based on it.
> 
> Please review my changes in the code and news file fragment.
> 
> Free free to revert any of my changes.

Your changes all looked great, thanks for just fixing some of those issues, nice to save the review cycle time.

Assigning back to adiroiban for comments on my two patched above / merging into the created svn branch.  Please let me know if this is not the correct thing to do at this point.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban removed owner|
|-|-|

Changes looks good. Latest patches are great :)

I am very happy with moving client connection into _UserAuth.

I am also happy with test_agentAuthentication

test_loseConnection is a bit odd... I tried to update the tests to include an explicit assertion.

Since I have also made some changes I think that this patch needs a final review from another person.

Putting it up for review.

Many thanks for your contribution!

**Comment by twisted-trac** [comment]
|<img alt="sarum90's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|sarum90 commented|
|-|-|

Thanks adiroiban for all your work on this as well.

I just noticed a typo of mine in one of the docstrings - above patch fixes it.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

I don't know why it was not raised by twisted checker.

I have pushed the fix and send to buildbot.

I will try to push for someone else to review the changes as I think that we are pretty close to have this merged.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set owner to @adiroiban|
|-|-|

This seems reasonable to me. Just some minor nitpicks:

- Could the new API in conch.endpoints have ```@since: 16.0``` added to their docstrings? We should really be doing this for new APIs, but we don't have an automated checker for it.

- I don't really understand the topfile. Maybe "twisted.conch.endpoint.SSHCommandClientEndpoint, when authentication is delegated to an agent, no longer leaves the agent connection opened when connection to the server is lost."

Please consider these two optional points, and land.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to sarum90|
|-|-|

Many thanks for the review.

This is private API and I don't see the point of adding the since reference. How does it help?

Glyph also commented this over IRC, but after finding out that this is private API, he was fine not to add it.

-----

I have updated the topfile. Thanks!

Will merge!

Many thanks to everyone! I am happy to see conch.ssh receiving more love :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46551]) Merge conch-endpoints-newconnection-agent-8138: Fix agent connection leak in twisted.conch.endpoints._NewConnectionHelper.

Author: sarum90, adiroiban
Reviewer: adiroiban, hawkowl
Fixes: [#8138](https://github.com/twisted/twisted/issues/8138)
